File: tools/snap.js
// Snapshot live p5.js sketches: the still that the PowerPoint and the PDF show where the html deck runs the sketch.
// usage: NODE_PATH=$(npm root -g) node tools/snap.js <page.html> <out.png> <canvas w> <canvas h> [more quadruples...]
// The still is taken 1.5 s after load with the mouse resting at 60 % / 40 % of the canvas, no click: design the
// sketch so that this resting state is a good picture. tools/snap.py drives this for whole decks.
const path = require('path');
const { chromium } = require('playwright');

(async () => {
  const args = process.argv.slice(2);
  const browser = await chromium.launch({ executablePath: process.env.CHROMIUM || undefined });
  let failed = 0;
  for (let i = 0; i + 3 < args.length; i += 4) {
    const [file, out] = [args[i], args[i + 1]];
    const cw = Number(args[i + 2]), ch = Number(args[i + 3]);
    const page = await browser.newPage({ viewport: { width: cw, height: ch }, deviceScaleFactor: 2 });
    page.on('pageerror', e => console.log(`error in ${path.basename(file)}: ${String(e).slice(0, 200)}`));
    page.on('console', m => { if (m.type() === 'error') console.log(`console in ${path.basename(file)}: ${m.text().slice(0, 200)}`); });
    try {
      await page.goto('file://' + path.resolve(file) + '?snap', { waitUntil: 'load' });
      await page.waitForSelector('canvas', { state: 'visible', timeout: 8000 }).catch(() => null);
      await page.mouse.move(Math.round(cw * 0.6), Math.round(ch * 0.4));
      await page.waitForTimeout(1500);
      const c = await page.$('canvas');
      if (c && await c.isVisible()) {
        await c.screenshot({ path: out, timeout: 10000 });
        console.log(`snap: ${path.basename(out)}`);
      } else {                   // p5 not loaded, or the sketch threw before createCanvas: no still, the old one stays
        console.log(`no canvas in ${path.basename(file)}: nothing written to ${path.basename(out)}`);
        failed += 1;
      }
    } catch (e) {                // one bad page must not stop the rest of the batch
      console.log(`failed ${path.basename(file)}: ${String(e).split('\n')[0].slice(0, 200)}; nothing written to ${path.basename(out)}`);
      failed += 1;
    }
    await page.close();
  }
  await browser.close();
  process.exitCode = failed ? 1 : 0;
})();

File: tools/snap.py
"""
Snapshot the live sketches of one or more decks into deck/assets/sketches/ (committed):
the still the PowerPoint and the PDF show where the html deck runs the sketch.

    python tools/snap.py week03            # the sketches of week 3 that have no snapshot yet
    python tools/snap.py week03 --force    # redo them
    python tools/snap.py --all [--force]   # every deck

Needs node + playwright + Chromium (see README); works on a clean checkout (the p5 the pages load is
copied into _site first). The still is taken 1.5 s after load with the mouse resting at 60 % / 40 % of
the canvas, no click; design the sketch so that state looks right. A page that makes no canvas writes
no still (the committed one stays) and the exit code is 1.
"""
from __future__ import annotations

import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT / 'tools'))

import deckgen                          # noqa: E402
from build_all import DECKS, load_deck  # noqa: E402


def main(argv):
    force = '--force' in argv
    names = [a for a in argv if not a.startswith('--')]
    if '--all' in argv or not names:
        names = DECKS
    if deckgen.node_playwright_env() is None:
        print('node + playwright + Chromium are needed for snapshots (README: Build locally)')
        return 1
    failed = 0
    for name in names:
        mod = load_deck(name)
        wanted = [el for el in deckgen.sketches_of(mod.DECK) if el.twin and (force or not deckgen.twin_png(el.name).exists())]
        made = deckgen.snapshot_sketches(mod.DECK, name, force=force)
        failed += len(wanted) - len(made)
        missing = [el.name for el in deckgen.sketches_of(mod.DECK) if el.twin and not deckgen.twin_png(el.name).exists()]
        print(f'{name}: {len(made)} snapshot(s) made' + (f', still missing: {missing}' if missing else ''))
    return 1 if failed else 0


if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))
